### DISCUSSION TITLE
Speeding up JSONL to Parquet Conversions

### DISCUSSION THREAD

**Discussion by marklit** [duckdb/duckdb#6478] [General]
Are there any user-facing settings I could use to speed up JSONL to Parquet conversions with DuckDB? I'm able to get DuckDB to produce PQs within 1.3x the speed of ClickHouse on an Intel-based 2020 MBP (albeit on an external SSD) but on the following GCP VM there is a 3.37x difference in performance.

I've been looking at a number of tools recently and on macOS, DuckDB is as close to ClickHouse in terms of performance that I can find. I'm wondering if compilation settings or some user setting could help it match or exceed ClickHouse. Last month I published a blog post on this albeit I didn't look at DuckDB at the time: https://tech.marksblogg.com/postgresql-to-bigquery.html

The following was run on Ubuntu 20 on a `e2-highcpu-32` GCP VM with 32 GB of RAM and 32 vCPUs.

I downloaded the California dataset from https://github.com/microsoft/USBuildingFootprints and converted it from JSONL into Parquet with DuckDB v0.7.0 f7827396d7 and ClickHouse 23.3.1.265.

```bash
$ ogr2ogr -f GeoJSONSeq /vsistdout/ California.geojson \
    | jq -c '.properties * {geom: .geometry|tostring}' \
    > California.jsonl
$ head -n1 California.jsonl | jq .
```

```json
{
  "release": 1,
  "capture_dates_range": "",
  "geom": "{\"type\":\"Polygon\",\"coordinates\":[[[-114.127454,34.265674],[-114.127476,34.265839],[-114.127588,34.265829],[-114.127565,34.265663],[-114.127454,34.265674]]]}"
}
```

ClickHouse produced a 794 MB Parquet file in 13.449s and DuckDB produced a 867 MB Parquet file in 45.372s.

```bash
$ time cat California.jsonl \
    | clickhouse local \
        --input-format JSONEachRow \
        -q "SELECT *
            FROM table
            FORMAT Parquet" \
    > cali.ch.pq # 13.449s, 794 MB

$ time echo "COPY (SELECT *
                   FROM read_json_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy',
                                  ROW_GROUP_SIZE 130099920);" \
    | ~/duckdb # 45.372s, 867 MB
```

The resulting DuckDB-produced Parquet matches ClickHouse in terms of row groups though the resulting file sizes do differ by 71 MB. I hard-coded the row group size as I can't set that with ClickHouse but I wanted DuckDB to match the PQ settings as much as I could.

```python
from collections import Counter
from operator    import itemgetter
from pprint      import pprint

import pyarrow.parquet as pq


pf = pq.ParquetFile('cali.duckdb.pq')
print(pf.metadata)
```

```python
<pyarrow._parquet.FileMetaData object at 0x7f56288f4180>
  created_by: DuckDB
  num_columns: 1
  num_rows: 11542912
  num_row_groups: 1
  format_version: 1.0
  serialized_size: 1599
```

```python
print(Counter(pf.metadata.row_group(rg).num_rows
              for rg in range(0, pf.metadata.num_row_groups)))
```

```python
Counter({11542912: 1})
```

```python
print(set(pf.metadata.row_group(rg).column(col).compression
          for col in range(0, pf.metadata.num_columns)
          for rg in range(0, pf.metadata.num_row_groups)))
```

```python
{'SNAPPY'}
```

```python
lowest_val, lowest_rg, lowest_col = None, None, None

for rg in range(0, pf.metadata.num_row_groups):
    for col in range(0, pf.metadata.num_columns):
        x = pf.metadata.row_group(rg).column(col)

        ratio = x.total_compressed_size / x.total_uncompressed_size

        if not lowest_val or lowest_val > ratio:
            lowest_val = ratio
            lowest_rg, lowest_col = rg, col

print(pf.metadata.row_group(lowest_rg).column(lowest_col))
```

```python
<pyarrow._parquet.ColumnChunkMetaData object at 0x7f55ca965ae0>
  file_offset: 0
  file_path:
  physical_type: BYTE_ARRAY
  num_values: 11542912
  path_in_schema: json
  is_stats_set: True
  statistics:
    <pyarrow._parquet.Statistics object at 0x7f55ca965db0>
      has_min_max: True
      min: {"release":1,"capture_dates_range":"","geom":"{\"type\":\"Polygon\",\"coordinates\":[[[-114.125395,34.261546],[-114.125465,34.261573],[-114.125506,34.261499],[-114.125436,34.261473],[-114.125395,34.261546]]]}"}
      max: {"release":2,"capture_dates_range":"9/27/2017-3/5/2020","geom":"{\"type\":\"Polygon\",\"coordinates\":[[[-119.496708,36.574252],[-119.496531,36.57425],[-119.496524,36.574659],[-119.496312,36.574657],[-119.496311,36.574722],[-119.496064,36.574719],[-119.496063,36.574814],[-119.496112,36.574814],[-119.496111,36.574864],[-119.49655,36.574869],[-119.496549,36.574914],[-119.496686,36.574915],[-119.496685,36.574933],[-119.496747,36.574933],[-119.496748,36.574874],[-119.496697,36.574874],[-119.496708,36.574252]]]}"}
      null_count: 0
      distinct_count: 0
      num_values: 11542912
      physical_type: BYTE_ARRAY
      logical_type: String
      converted_type (legacy): UTF8
  compression: SNAPPY
  encodings: ('PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN', 'PLAIN')
  has_dictionary_page: False
  dictionary_page_offset: None
  data_page_offset: 4
  total_compressed_size: 908551892
  total_uncompressed_size: 3240748574
```

```python
ratios = []

for rg in range(0, pf.metadata.num_row_groups):
    for col in range(0, pf.metadata.num_columns):
        x = pf.metadata.row_group(rg).column(col)
        ratio = x.total_compressed_size / x.total_uncompressed_size
        ratios.append('%.1f' % ratio)

pprint(sorted(Counter(ratios).items(),
              key=itemgetter(0)))
```

```python
[('0.3', 1)]
```

Below is a Flamegraph of DuckDB's execution.

![parquet duckdb snappy](https://user-images.githubusercontent.com/359316/221520525-b4674c07-e2fc-4f01-aae2-2e923e83c628.svg)

I ran a 10-line version of the above file through DuckDB. This is what `strace` and `perf` reported.

```bash
$ vi create.sql
```

```sql
COPY (SELECT *
      FROM read_json_objects('/home/mark/cali10.jsonl'))
TO '/home/mark/cali10.duckdb.pq' (FORMAT 'PARQUET',
                                CODEC  'Snappy',
                                ROW_GROUP_SIZE 130099920);
```

```bash
$ sudo su
$ strace -wc /home/mark/duckdb < /home/mark/create.sql
```

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 18.78    0.002675        2674         1           fsync
 16.83    0.002397          24        96           mmap
 12.62    0.001797          57        31           clone
 11.42    0.001626          40        40           futex
  9.15    0.001304          34        38           mprotect
  7.16    0.001019          30        33           munmap
  2.88    0.000410          27        15         2 openat
  2.68    0.000381          25        15           close
  2.51    0.000357          25        14           brk
  2.39    0.000340          26        13           read
  1.57    0.000223          13        16           fstat
  1.47    0.000209         209         1           execve
  1.35    0.000192          21         9           pread64
  1.07    0.000152          76         2         2 connect
  0.94    0.000134          44         3           rt_sigaction
  0.91    0.000130         129         1           write
  0.91    0.000129          64         2         1 ioctl
  0.91    0.000129          32         4         3 access
  0.88    0.000126          62         2           socket
  0.73    0.000104          51         2         1 readlink
  0.60    0.000086          42         2           lseek
  0.32    0.000046          46         1           set_tid_address
  0.32    0.000045          45         1           set_robust_list
  0.31    0.000044          44         1           prlimit64
  0.29    0.000041          40         1           sysinfo
  0.28    0.000040          40         1           getuid
  0.28    0.000040          39         1           rt_sigprocmask
  0.16    0.000023          22         1           fcntl
  0.14    0.000020           9         2         1 arch_prctl
  0.09    0.000012          12         1           stat
  0.08    0.000011          11         1           sched_getaffinity
------ ----------- ----------- --------- --------- ----------------
100.00    0.014242                   351        10 total
```

```bash
$ perf stat -dd /home/mark/duckdb < /home/mark/create.sql
```

```
 Performance counter stats for '/home/mark/duckdb':

             22.88 msec task-clock                #    1.169 CPUs utilized
               170      context-switches          #    7.429 K/sec
                69      cpu-migrations            #    3.015 K/sec
               925      page-faults               #   40.422 K/sec
   <not supported>      cycles
   <not supported>      instructions
   <not supported>      branches
   <not supported>      branch-misses
   <not supported>      L1-dcache-loads
   <not supported>      L1-dcache-load-misses
   <not supported>      LLC-loads
   <not supported>      LLC-load-misses
   <not supported>      L1-icache-loads
   <not supported>      L1-icache-load-misses
   <not supported>      dTLB-loads
   <not supported>      dTLB-load-misses
   <not supported>      iTLB-loads
   <not supported>      iTLB-load-misses

       0.019567825 seconds time elapsed

       0.006214000 seconds user
       0.018630000 seconds sys
```

**Comment by hannes** [comment]
There are two options you could try to speed up Parquet writing:

- Enable per-thread output:
     `COPY (...) TO ... (FORMAT PARQUET, PER_THREAD_OUTPUT TRUE);`
- Disable insertion order preservation:
 `SET preserve_insertion_order=false;`

**Reply by marklit** [reply]
Great, I'll give those two a try.

**Comment by marklit** [comment]
The following finished in 11m25s on the MBP I've been running these tests on.

```bash
$ time echo "PRAGMA temp_directory='/Volumes/Seagate/open5g_data/microsoft_buildings';
            SET preserve_insertion_order=false;
            COPY (SELECT *
                 FROM read_json_objects('California.jsonl'))
            TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                 CODEC  'Snappy',
                                 ROW_GROUP_SIZE 130099920,
                                 PER_THREAD_OUTPUT TRUE);" \
    | ~/Downloads/duckdb
```

Without the PRAGMA statement, I got `Out of Memory Error: failed to allocate data of size 2147483648` returned. I set the temp folder to the same as the data folder on my external SSD.

Looking at the Flamegraph, about 34% of the time is spent in `duckdb::StringColumnWriter::Analyze` under `duckdb::ParquetWriter::Flush`, the dataset is mostly made up of strings, is type inference a bottleneck or am I reading this wrong?

**Reply by marklit** [reply]
Setting the temp folder to the data folder was a bad idea as it wiped out the data folder after. I can recover the files from elsewhere.

**Reply by Tishj** [reply]
The `2147483648` looks suspicious, as that's int32_t max

**Reply by Tishj** [reply]
Also I think the issue you mentioned about wiping out the temp folder is fixed by a recent PR :)

**Reply by Mytherin** [reply]
That is related to the string dictionary construction. I would recommend writing smaller row groups. DuckDB defaults to 120K rows per row group for good reason - I would not recommend writing 130M rows per row group as it will require the system to buffer 130M rows per thread, and causes the system to run dictionary analysis over 130M strings at a time (which it is not specifically designed for).

**Reply by lnkuiper** [reply]
~~I'm also a bit worried about the setting `ROW_GROUP_SIZE 130099920`. The idea of row groups is that readscan be parallelized across each row group. We tend so set this value around 100k. However, by setting the value to 130 million, I think that the entire parquet file will be exactly one row group, which means that reading the file cannot be done in parallel.~~

Mark beat me to it :)

**Comment by marklit** [comment]
Getting closer..

This ran in 30.673s (ClickHouse was 35.243s):

```bash
$ time echo "SET preserve_insertion_order=false;
             COPY (SELECT *
                 FROM read_json_objects('California.jsonl'))
           TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                CODEC  'Snappy',
                                PER_THREAD_OUTPUT TRUE);" \
    | ~/Downloads/duckdb
```

This ran in 31.184s:

```bash
$ time echo "SET preserve_insertion_order=false;
             COPY (SELECT *
                 FROM read_json_objects('California.jsonl'))
           TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                CODEC  'Snappy');" \
    | ~/Downloads/duckdb
```

This ran in 31.022s:

```bash
$ time echo "COPY (SELECT *
                 FROM read_json_objects('California.jsonl'))
           TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                CODEC  'Snappy',
                                PER_THREAD_OUTPUT TRUE);" \
    | ~/Downloads/duckdb
```

This ran in 30.711s:

```bash
$ time echo "COPY (SELECT *
                 FROM read_json_objects('California.jsonl'))
           TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                CODEC  'Snappy');" \
    | ~/Downloads/duckdb
```

**Reply by lnkuiper** [reply]
Could you try using the `read_ndjson_objects` function instead? With `PER_THREAD_OUTPUT TRUE`.
I would expect this to be significantly faster

**Comment by marklit** [comment]
25.146s, not bad. 868 MB across 8 files.

```bash
$ time echo "COPY (SELECT *
                   FROM read_ndjson_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy',
                                  PER_THREAD_OUTPUT TRUE);" \
      | ~/Downloads/duckdb
```

30.826s if I produce a single PQ file.

```bash
$ time echo "COPY (SELECT *
                   FROM read_ndjson_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy');" \
      | ~/Downloads/duckdb
```

Any other ideas before I launch another GCP VM?

**Reply by lnkuiper** [reply]
I think that's all of my ideas. If you are interested in decomposing the JSON, i.e., converting the key/values in the JSON to columnar data, you could try `read_json_auto`. This kind of JSON:
```json
{
  "release": 1,
  "capture_dates_range": "",
  "geom": "{\"type\":\"Polygon\",\"coordinates\":[[[-114.127454,34.265674],[-114.127476,34.265839],[-114.127588,34.265829],[-114.127565,34.265663],[-114.127454,34.265674]]]}"
}
```
would be converted into a table with 3 columns instead. The `"coordinates"` field would become a DuckDB list of doubles, rather than a JSON string. By decomposing the data we get rid of the redundancy of JSON, i.e., the name of every column showing up in every record. I would expect the resulting parquet file to be much smaller since it is no longer full of strings.

However, I realise that GeoJSON is a format of it's own, and that perhaps you prefer to keep it intact. Up to you!

**Reply by marklit** [reply]
In my normal workloads, I keep WKBs or GeoJSON as they are so this test is as close to typical as I can get with a public dataset.

**Comment by hannes** [comment]
I think that should get you good parallelism, let us know how it goes!

**Comment by marklit** [comment]
Much closer, I launched the same VM type and they're now within 1.54x of one another.


```bash
$ time cat California.jsonl \
    | clickhouse local \
        --input-format JSONEachRow \
        -q "SELECT *
            FROM table
            FORMAT Parquet" \
    > cali.ch.pq # 12.747s, 794 MB

$ time echo "COPY (SELECT *
                   FROM read_ndjson_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy',
                                  PER_THREAD_OUTPUT TRUE);" \
    | ~/duckdb # 19.664s, 867 MB
```

On the 32-core system, `htop` reports both DuckDB and CH maxing out at 200%. I'm wondering if there is a way to get more cores involved with this workload.

I can also produce a single file in 36.822s with `read_ndjson_objects` which is a step-up from what `read_json_objects` was doing.

```bash
$ time echo "COPY (SELECT *
                   FROM read_ndjson_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy');" \
    | ~/duckdb # 36.822s, 867 MB
```

**Reply by Mytherin** [reply]
Perhaps give `SET preserve_insertion_order=false` a try as well, without the `PER_THREAD_OUTPUT` flag?

**Comment by marklit** [comment]
The following came back in 18.461s on the GCP VM:

```bash
$ time echo "SET preserve_insertion_order=false;
             COPY (SELECT *
                   FROM read_ndjson_objects('California.jsonl'))
             TO 'cali.duckdb.pq' (FORMAT 'PARQUET',
                                  CODEC  'Snappy',
                                  PER_THREAD_OUTPUT TRUE);" \
    | ~/duckdb
```

**Reply by hannes** [reply]
Sounds good, what was the ClickHouse time on that VM?

**Reply by marklit** [reply]
12.747s

**Reply by marklit** [reply]
Do you think there are any compilation changes that could be made to bring DuckDB closer to 12.747s? We were able to get DuckDB running faster than CH on macOS. Surely there is something that could speed it up a bit more on Linux.

**Reply by Mytherin** [reply]
Looking at the profiling info it seems like a lot of time is going into analyzing for dictionary compression - we use the `unordered_map` implementation of the system for that, which might have different performance characteristics based on which standard library is used.

In general this step is not useful and could be eliminated in this case, anyway. We should probably not attempt to do it for `JSON` columns to begin with or have an early out in case there are many unique strings in a given column.

**Reply by lnkuiper** [reply]
We would have to do some profiling to see what is going on to be able to speed it up on Linux.

One thing we discussed is that our Parquet writer tries to do dictionary compression on string columns. However, in this case the string column is JSON, and there is no repetition, so this is a waste of time. This would probably be the largest gain.

Finally, if both DuckDB and ClickHouse max out at 200%, then the disk of the machine you're using might be causing the bottleneck, but I'm not sure.

Edit: Mark beat me to it again! Was writing this reply at the same time as Mark
